### ISSUE TITLE
the xml for validation KEY not is the correct

### ISSUE THREAD

**Issue by eriverac** [NVIDIA/libnvidia-container#70]
I'm tray to install nvidia-docker in Red Hat 7.6 on ppc64lc, the repository not have the key correct
I make the repository whit the script

DIST=$(. /etc/os-release; echo $ID$VERSION_ID)
curl -s -L https://nvidia.github.io/libnvidia-container/$DIST/libnvidia-container.repo | \
  sudo tee /etc/yum.repos.d/libnvidia-container.repo

[root@mia01 7Server]# DIST=$(. /etc/os-release; echo $ID$VERSION_ID)
[root@mia01 7Server]# curl -s -L https://nvidia.github.io/libnvidia-container/$DIST/libnvidia-container.repo | \
>   sudo tee /etc/yum.repos.d/libnvidia-container.repo
# Unsupported distribution!
# Check https://nvidia.github.io/libnvidia-container

**Comment by DabeDotCom** [comment]
I'm not trying to hijack this thread, but around the same time, I started getting ***repomd.xml could not be verified*** errors, as well...

For me, though, I **can** download `/etc/yum.repos.d/libnvidia-container.repo`.  Once in place, however, `sudo yum update` *et al* fail with:

```
Loaded plugins: fastestmirror
Loading mirror speeds from cached hostfile
epel/x86_64/metalink                                                                                                                |  16 kB  00:00:00     
 * base: mirror.vtti.vt.edu
 * epel: mirror.umd.edu
 * extras: mirror.es.its.nyu.edu
 * rpmfusion-free-updates: mirror.math.princeton.edu
 * updates: mirror.ash.fastserv.com

libnvidia-container/x86_64/signature                                          |  455 B  00:00:00     
Retrieving key from https://nvidia.github.io/libnvidia-container/gpgkey
libnvidia-container/x86_64/signature                                          | 2.0 kB  00:00:00 !!! 
https://nvidia.github.io/libnvidia-container/centos7/x86_64/repodata/repomd.xml: [Errno -1] repomd.xml signature could not be verified for libnvidia-container
Trying other mirror.

 One of the configured repositories failed (libnvidia-container),
 and yum doesn't have enough cached data to continue. At this point the only
 safe thing yum can do is fail. There are a few ways to work "fix" this:

     1. Contact the upstream for the repository and get them to fix the problem.

     2. Reconfigure the baseurl/etc. for the repository, to point to a working
        upstream. This is most often useful if you are using a newer
        distribution release than is supported by the repository (and the
        packages for the previous distribution release still work).
       [...]

failure: repodata/repomd.xml from libnvidia-container: [Errno 256] No more mirrors to try.
https://nvidia.github.io/libnvidia-container/centos7/x86_64/repodata/repomd.xml: [Errno -1] repomd.xml signature could not be verified for libnvidia-container
```

Interestingly, I *can* verify that `repomd.xml` **DOES** seem to be signed correctly — at least according to GPG:
```
mkdir /tmp/nvidia-container
cd /tmp/nvidia-container
baseurl="https://nvidia.github.io/libnvidia-container"
repourl="${baseurl}/centos7/x86_64/repodata"

### Fetch GPG Public Key
#
#  NOTE: "gpg --keyring" requires us to remove the ASCII Armor

curl -s -L -O "${baseurl}/gpgkey"
gpg --dearmor gpgkey

### "gpgkey.gpg" now contains the GPG Public Key in binary form...
#
#  prompt% gpg gpgkey.gpg
#  pub  4096R/F796ECB0 2017-09-28 NVIDIA CORPORATION (Open Source Projects) <cudatools@nvidia.com>
#  sub  2048R/A04EA552 2017-09-28 [expires: 2018-09-28]
#  sub  2048R/AC1160CD 2019-09-18 [expires: 2020-09-17]

### Fetch "repomd.xml" and its GPG Signature
#
#  NOTE: Yes, ideally these would come from a different source than above...

curl -s -L -O "${repourl}/repomd.xml"
curl -s -L -O "${repourl}/repomd.xml.asc"

### Finally, Verify 'repomd.xml's Signature
#
#  NOTE: "--keyring" has to include a directory path, else it defaults to "~/.gnupg/"

gpg --no-default-keyring --keyring ./gpgkey.gpg --verify repomd.xml.asc repomd.xml
```

This outputs:
```
gpg: Signature made Fri 20 Sep 2019 04:58:46 PM EDT using RSA key ID AC1160CD
gpg: Good signature from "NVIDIA CORPORATION (Open Source Projects) <cudatools@nvidia.com>"
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: C95B 321B 61E8 8C18 09C4  F759 DDCA E044 F796 ECB0
     Subkey fingerprint: 2041 629B F535 2FAC 178A  9B1F 6ED9 1CA3 AC11 60CD
```

So, yeah, "Good signature" ... "using key ID AC1160CD".  T'sall Good!  Yet `sudo yum check-update` is still *no bueno*.  :-(

Anyway, I've tried clearing my yum cache, `rpm --import`-ing the gpgkey, even saving it in `file:///etc/pki/rpm-gpg/` and pointing my `yum.repos.d` at *that*, but to no avail!  «sigh»  I feel like I must be missing something...  Maybe `yum` 

PS — I'm not interested in, "Just set `repo_gpgcheck = 0`" or "Run with `--nogpgcheck`" vulnerabilities...  The goal isn't to make it work badly; the goal is to make it work WELL.


# UPDATE

I tracked down the issue to the following section of `/usr/lib/python2.7/site-packages/yum/misc.py`
```
def valid_detached_sig(sig_file, signed_file, gpghome=None):
    """takes signature , file that was signed and an optional gpghomedir"""

# ...
    ctx = gpgme.Context()

    try:
        sigs = ctx.verify(sig, signed_text, plaintext)
    except gpgme.GpgmeError, e:
        return False
    else:
        if not sigs:
            return False
        # is there ever a case where we care about a sig beyond the first one?
        thissig = sigs[0]
        if not thissig:
            return False

        if thissig.validity in (gpgme.VALIDITY_FULL, gpgme.VALIDITY_MARGINAL,
                                gpgme.VALIDITY_ULTIMATE):
            return True

    return False
```

***AHA!***  Apparently whoever refreshed the GPG Public Key left the old, expired subkey in place...  But `yum` doesn't actually *check* any beyond the first!

I'll look into submitting a patch to Redhat to fix their "premature optimization", but in the meantime, if it's not too much effort, it would serve the community well if somebody at NVIDIA could re-export that `gpgkey` file ***without*** the expired `2048R/A04EA552` subkey.

THANKS!!  :-D

**Comment by RenaudWasTaken** [comment]
Let me do that, sorry about the churn :)
Also adding the missing rhel7.6 support

**Comment by klueska** [comment]
This is the support matrix for the recently released 1.1.0 packages:
```
+----------------------+-------------+----------------+---------+-----------------+
|  OS Name / Version   |  Identifier | amd64 / x86_64 | ppc64le | arm64 / aarch64 |
+======================+=============+================+=========+=================+
| Amazon Linux 1       | amzn1       |       X        |         |                 |
| Amazon Linux 2       | amzn2       |       X        |         |                 |
| Amazon Linux 2017.09 | amzn2017.09 |       X        |         |                 |
| Amazon Linux 2018.03 | amzn2018.03 |       X        |         |                 |
| Open Suse Leap 15.0  | sles15.0    |       X        |         |                 |
| Open Suse Leap 15.1  | sles15.1    |       X        |         |                 |
| Debian Linux 9       | debian9     |       X        |         |                 |
| Debian Linux 10      | debian10    |       X        |         |                 |
| Centos 7             | centos7     |       X        |    X    |                 |
| Centos 8             | centos8     |       X        |    X    |        X        |
| RHEL 7.4             | rhel7.4     |       X        |    X    |                 |
| RHEL 7.5             | rhel7.5     |       X        |    X    |                 |
| RHEL 7.6             | rhel7.6     |       X        |    X    |                 |
| RHEL 7.7             | rhel7.7     |       X        |    X    |                 |
| RHEL 8.0             | rhel8.0     |       X        |    X    |        X        |
| RHEL 8.1             | rhel8.1     |       X        |    X    |        X        |
| RHEL 8.2             | rhel8.2     |       X        |    X    |        X        |
| Ubuntu 16.04         | ubuntu16.04 |       X        |    X    |                 |
| Ubuntu 18.04         | ubuntu18.04 |       X        |    X    |        X        |
| Ubuntu 19.04         | ubuntu19.04 |       X        |    X    |        X        |
| Ubuntu 19.10         | ubuntu19.10 |       X        |    X    |        X        |
| Ubuntu 20.04         | ubuntu20.04 |       X        |    X    |        X        |
+----------------------+-------------+----------------+---------+-----------------+
```

Please let us know if this resolves your issue.
